extensions/sql-database-projects/src/dialogs/deployDatabaseDialog.ts

```diff
@@ -0,0 +1,291 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as azdata from 'azdata';
+import * as constants from '../common/constants';
+import * as path from 'path';
+import { Project } from '../models/project';
+import { DataSource } from '../models/dataSources/dataSources';
+import { SqlConnectionDataSource } from '../models/dataSources/sqlConnectionStringSource';
+
+interface DataSourceDropdownValue extends azdata.CategoryValue {
+	dataSource: DataSource;
+	database: string;
+}
+
+export class DeployDatabaseDialog {
+	public dialog: azdata.window.Dialog;
+	public deployTab: azdata.window.DialogTab;
+	private targetConnectionTextBox: azdata.InputBoxComponent | undefined;
+	private targetConnectionFormComponent: azdata.FormComponent | undefined;
+	private dataSourcesFormComponent: azdata.FormComponent | undefined;
+	private dataSourcesDropDown: azdata.DropDownComponent | undefined;
+	private targetDatabaseTextBox: azdata.InputBoxComponent | undefined;
+	private deployScriptNameTextBox: azdata.InputBoxComponent | undefined;
+	private connectionsRadioButton: azdata.RadioButtonComponent | undefined;
+	private dataSourcesRadioButton: azdata.RadioButtonComponent | undefined;
+	private formBuilder: azdata.FormBuilder | undefined;
+
+	private connection: azdata.connection.ConnectionProfile | undefined;
+	private connectionIsDataSource: boolean | undefined;
+
+	constructor(private project: Project) {
+		this.dialog = azdata.window.createModelViewDialog(constants.deployDialogName);
+		this.deployTab = azdata.window.createTab(constants.deployDialogName);
+	}
+
+	public async openDialog(): Promise<void> {
+		this.initializeDialog();
+		this.dialog.okButton.label = constants.deployDialogOkButtonText;
+		this.dialog.okButton.enabled = false;
+		this.dialog.okButton.onClick(async () => await this.deploy());
+
+		this.dialog.cancelButton.label = constants.cancelButtonText;
+		this.dialog.cancelButton.onClick(async () => await this.cancel());
+		this.dialog.cancelButton.onClick(async () => await this.cancel());
+
+		let generateScriptButton: azdata.window.Button = azdata.window.createButton(constants.generateScriptButtonText);
+		generateScriptButton.onClick(async () => await this.generateScript());
+		generateScriptButton.enabled = false;
+
+		this.dialog.customButtons = [];
+		this.dialog.customButtons.push(generateScriptButton);
+
+		azdata.window.openDialog(this.dialog);
+	}
+
+
+	private initializeDialog(): void {
+		this.initializeDeployTab();
+		this.dialog.content = [this.deployTab];
+	}
+
+	private initializeDeployTab(): void {
```

Comment:
This method returns instantly, but isn't actually "complete" when the return is done.

You might want to wrap it around a promise.